File: leetcode/328_odd_even_linked_list.py
class ListNode:

    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:

    def oddEvenList(self, head: ListNode) -> ListNode:
        if not head:
            return head
        event_start = head.next
        curr = odd_end = head
        index = 1
        while curr and curr.next:
            if index & 1:
                odd_end = curr.next.next or curr
            index += 1
            next_node = curr.next
            curr.next = curr.next.next
            curr = next_node
        odd_end.next = event_start
        return head

File: leetcode/java/328OddEvenLinkedList.java
/**
 * T: O(N) S: O(1)
 *
 * <p>First, we check if we can advance further. If so - next odd is after the current even, after
 * that - the new even is after the updated odd. We have to check for the even in the condition so
 * that we don't advance the latest odd to null, which will fail the last step - joining last odd to
 * the first even.
 */
public class ListNode {
  int val;
  ListNode next;

  ListNode() {}

  ListNode(int val) {
    this.val = val;
  }

  ListNode(int val, ListNode next) {
    this.val = val;
    this.next = next;
  }
}

class Solution {
  public ListNode oddEvenList(ListNode head) {
    if (head == null) {
      return null;
    }
    ListNode firstEven = head.next;
    ListNode currOdd = head;
    ListNode currEven = firstEven;
    while (currEven != null && currEven.next != null) {
      currOdd.next = currEven.next;
      currOdd = currOdd.next;
      currEven.next = currOdd.next;
      currEven = currEven.next;
    }
    currOdd.next = firstEven;
    return head;
  }
}
